eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientInstantiationCheck.java

```diff
@@ -0,0 +1,167 @@
+                break;
+            case TokenTypes.LITERAL_CLASS:
+                checkServiceClientNaming(token);
+                break;
+            case TokenTypes.OBJBLOCK:
+                checkClassFieldFinal(token);
+                break;
+            default:
+                // Checkstyle complains if there's no default block in switch
+                break;
+        }
+    }
+
+    /**
+     *  Checks if the class is annotated with @ServiceClient annotation
+     *
+     * @param token the CLASS_DEF AST node
+     * @return true if the class is annotated with @ServiceClient, false otherwise.
+     */
+    private boolean hasServiceClientAnnotation(DetailAST token) {
+        DetailAST modifiersToken = token.findFirstToken(TokenTypes.MODIFIERS);
+        if (modifiersToken != null) {
+            for (DetailAST ast = modifiersToken.getFirstChild(); ast != null; ast = ast.getNextSibling()) {
+                if (ast.getType() == TokenTypes.ANNOTATION) {
+                    DetailAST annotationIdent = ast.findFirstToken(TokenTypes.IDENT);
+                    if (annotationIdent != null && SERVICE_CLIENT.equals(annotationIdent.getText())) {
+                        return true;
+                    }
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     *  Checks for public or protected constructor for the service client class.
+     *  Log error if the service client has public or protected constructor.
+     *
+     * @param token the CTOR_DEF AST node
+     */
+    private void checkNonPublicOrProtectedConstructor(DetailAST token) {
+        DetailAST modifiersToken = token.findFirstToken(TokenTypes.MODIFIERS);
+        AccessModifier accessModifier = CheckUtil.getAccessModifierFromModifiersToken(modifiersToken);
+        if (accessModifier.equals(AccessModifier.PUBLIC) || accessModifier.equals(AccessModifier.PROTECTED)) {
+            log(modifiersToken, "Subclass of ServiceClient should not have any public or protected constructor.");
+        }
+    }
+
+    /**aa
+     * Checks for public static method named 'builder'. Should avoid to use method name, 'builder'.
+     *
+     * @param token the METHOD_DEF AST node
+     */
+    private void checkNonPublicStaticBuilderMethodName(DetailAST token) {
+        DetailAST methodNameToken = token.findFirstToken(TokenTypes.IDENT);
+        if (methodNameToken != null && BUILDER.equals(methodNameToken.getText())) {
+            DetailAST modifiersToken = token.findFirstToken(TokenTypes.MODIFIERS);
+            AccessModifier accessModifier = CheckUtil.getAccessModifierFromModifiersToken(modifiersToken);
+            if (accessModifier.equals(AccessModifier.PUBLIC)
+                && modifiersToken.branchContains(TokenTypes.LITERAL_STATIC)) {
+                log(modifiersToken, "Subclass of ServiceClient should not have a public static method named ''builder''.");
+            }
+        }
```

Comment:
Get into the habit of inverting these checks as Connie noted. Rather than get more and more nested, check the opposite and return.